### PR TITLE
Angular template support for Storybook

### PR DESCRIPTION
**Pull request by alterx**
Issue: https://github.com/storybooks/storybook/issues/2645

## What I did
Initial work to enable writing stories as follows:
```
storiesOf('Button', module)
  .add('with text', () => ({
    moduleMetadata: {
      entryComponents: [Button],
      declarations: [Button],
    },
    props: {
      text: 'Hello Button',
      onClick: (event) => {
        console.log('some bindings work');
        console.log(event)
      }
    },
    template: `
      <h1> This is a template </h1>
      <button-component [text]="text" (onClick)="onClick($event)"></button-component>
    `
  }))
```

## How to test
```
yarn bootstrap --reset --core
cd examples/angular-cli
yarn storybook
open http://localhost:9009
```

### CODE REVIEW

**Comment by amcdnl** [comment]
@alterx - I don't think `template` and `component` should be used at the same time.

**Review by amcdnl — APPROVED** [review]
LGTM

**Comment by alterx** [comment]
@amcdnl we currently need the component for:
a) Adding it to our NgModule
b) Getting the metadata

Even if we make the user provide the bindings as a separate object we'd still need the component to add it to our inner module.

**Comment by igor-dv** [comment]
@alterx , what if a template contains few components? Maybe worth adding them to the `moduleMetadata` ?

**Comment by alterx** [comment]
@igor-dv yeah, in that case we would add them to the`moduleMetadata` (it does it like that right now) but the part that worries me (and it's not yet resolved) is that we also need the actual metadata from each component so we can provide the prop. 

This is an initial shot at the issue, I'd like @amcdnl to provide a real use case to see how we can implement these details

**Comment by amcdnl** [comment]
@alterx - I think you mis-understood me. I mean't for this example:

```
storiesOf('Button', module)
  .add('with text', () => ({
    moduleMetadata: {
       imports: [ButtonModule]
    },
    props: {
      text: 'Hello Button',
      onClick: (event) => {
        console.log('some bindings work');
        console.log(event)
      }
    },
    template: `
      <h1> This is a template </h1>
      <button-component [text]="text" (onClick)="onClick($event)"></button-component>
    `
  }))
```

Which makes it super confusing what `component` is actually for now.

**Comment by alterx** [comment]
@amcdnl yeah, I know you're talking about that example, I didn't misunderstood you. The component is still there because, currently, I'm using it to get the metadata.

**Comment by alterx** [comment]
This is just a contrived example, a POC. So it currently only supports creating the template using the bindings of that component we provide. That's why it has the label `do-not-merge` is not ready for primetime. I just submitted the PR for visibility (let people know is being worked on) and, of course, for you to test it.

**Comment by amcdnl** [comment]
@alterx - Why do you need it? I'm confused on that part I guess.

**Comment by alterx** [comment]
Because I didn't change the way this works. I just added the template parts on top of it so it still uses the component for props metadata. Hopefully it's not gonna be like that at the end. I just wanted to get it rendering a template.

**Comment by alterx** [comment]
With the use of `moduleMetadata` and some other changes this PR has also reduced the amount of code needed to bootstrap a component. The angular app is no longer doing manual checks on the metadata and only creates a dynamic component if the story provides a template*. The rest of the work is offloaded to Angular itself. Been wanting to do this for a while :D 
As a result, the API remains almost intact (I added the `template` option) while the code complexity has been reduced.

* Manual metadata checks are still needed inside the knobs addon (which needs to manually deal with the change events that come from the knobs channel and send them to the angular component).

File: app/angular/src/client/preview/angular/helpers.ts [hunk 1]
```diff
@@ -50,127 +44,75 @@ const debounce = (
...
+  return NgModule(moduleMeta)(moduleClass);
+};
 
-  const NewModule: any = function() {};
-  (<IModule>NewModule).annotations = [moduleMeta];
-  return NewModule;
+const DYNAMIC_COMPONENT_SELECTOR = 'storybook-dynamic-component';
+const createComponentFromTemplate = (
+  template: string,
+  params?: any[],
```
Comment (igor-dv) [inline app/angular/src/client/preview/angular/helpers.ts]: params are not used here

File: app/angular/src/client/preview/angular/helpers.ts [hunk 2]
```diff
@@ -1,29 +1,23 @@
-import { Type, enableProdMode, NgModule, Component, NgModuleRef } from '@angular/core';
+import { enableProdMode, NgModule, Component, NgModuleRef, Type } from '@angular/core';
 import { FormsModule } from '@angular/forms';
 import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 import { BrowserModule } from '@angular/platform-browser';
 import { AppComponent } from './components/app.component';
 import { ErrorComponent } from './components/error.component';
 import { NoPreviewComponent } from './components/no-preview.component';
 import { STORY } from './app.token';
-import { getAnnotations, getParameters, getPropMetadata } from './utils';
```
Comment (igor-dv) [inline app/angular/src/client/preview/angular/helpers.ts]: So, utils.ts can be deleted?

**Review by igor-dv — COMMENTED** [review]
This is so much easier and cleaner! Did you check it with v4?

**Comment by alterx** [comment]
Yup, checked in v4 :)

File: app/angular/src/client/preview/angular/helpers.ts [hunk 3]
```diff
@@ -134,23 +118,54 @@ const getModule = (
...
+    component,
+    componentMeta,
+    props,
+    propsMeta,
+    params,
+    moduleMeta,
+    template,
+  } = getComponentMetadata(currentStory(context));
 
   if (!componentMeta) {
```
Comment (amcdnl) [inline app/angular/src/client/preview/angular/helpers.ts]: I'd add some errors like:

```
  if (!template && (!component || typeof component !== 'function')) {
    throw new Error('No valid component provided');
  }

  if (template && component) {
    throw new Error('Component and template are mutually exclusive.');
  }
```

**Comment by teebszet** [comment]
Should there be documentation on this feature in https://storybook.js.org/basics/guide-angular/#write-your-stories?

**Comment by shilman** [comment]
@teebszet Any chance you can contribute some docs on this?

**Comment by breadadams** [comment]
Docs would be real handy, like Vue for example there are 2 ways of creating a story, whereas Angular shows the same twice.

Once I can wrap my head around this myself I'll be happy to update it 👍

**Pull request merged**